datadog/resource_datadog_authn_mapping.go

```diff
@@ -0,0 +1,203 @@
+package datadog
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/DataDog/datadog-api-client-go/api/v2/datadog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/terraform-providers/terraform-provider-datadog/datadog/internal/utils"
+)
+
+func resourceDatadogAuthnMapping() *schema.Resource {
+	return &schema.Resource{
+		Description:   "Provides a Datadog AuthN Mappings resource. This can be used to create and manage Datadog SAML Mappings.",
+		CreateContext: resourceDatadogAuthnMappingCreate,
+		ReadContext:   resourceDatadogAuthnMappingRead,
+		UpdateContext: resourceDatadogAuthnMappingUpdate,
+		DeleteContext: resourceDatadogAuthnMappingDelete,
+
+		Schema: map[string]*schema.Schema{
+			"key": {
+				Description: "Identity provider key.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"value": {
+				Description: "Identity provider value.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+			"role": {
+				Description: "The role to assign for key:value mapping.",
+				Type:        schema.TypeString,
+				Required:    true,
+			},
+		},
+	}
+}
+
+func resourceDatadogAuthnMappingCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```

Comment:
I think we're doing Terraform "wrong" by ignoring the ctx variable. The context can contain a cancellation handle and timeout, but I do see that most or all of this repo also ignores ctx because we're already using a separate context to encode the authentication credentials.

So, it looks like a much bigger problem than this PR, so I'm fine with keeping it as-is.

Reply:
Thanks for pointing this out. I didn't know about this. So does the correct way of using context involve binding a retry, cancel and timeout handler to be invoked?

Reply:
Yes and by putting the API & App key into an existing context, we force ourselves to ignore the context provided by Terraform, so Terraform can't cancel or timeout the request to the Datadog provider.

Instead of providing a common, shared context with credentials baked in, I think it would be better to have a helper method like this:
```
func addCredentialsToContext(ctx context.Context) context.Context {
     // takes an existing context (from Terraform)
     // and adds properties to it like Datadog credentials
     // so it can be used in the Datadog Go client
}
```

In this particular API, the API calls are (usually) fast so it's unlikely we'd hit timeouts or cancellations, so don't worry about it for this PR. But maybe we can revisit as we get more experience with Terraform.